### ISSUE TITLE
Add integration test for accessing "host.minikube.internal" in a container 

### ISSUE THREAD

**Issue by valters** [kubernetes/minikube#8439]
Hi,

The subtitle of article is "How to access host resources from a pod" but as far as I can tell, it does not demonstrate how to actually do it from inside a pod? For me it is unclear how to get the `host.minikube.internal` IP from inside a pod actually.

(My scenario is I run a Postgres install on local machine (bound to 0.0.0.0:5432) and would like a service from inside pod to connect to it.)

I can't figure out how to resolve `host.minikube.internal` from inside pod.

Here's an example:
```
$ minikube start --memory=16384 --cpus=4
😄  minikube v1.11.0 on Ubuntu 20.04
    ▪ MINIKUBE_ACTIVE_DOCKERD=minikube
✨  Using the kvm2 driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🔥  Creating kvm2 VM (CPUs=4, Memory=16384MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.18.3 on Docker 19.03.8 ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube"
$ minikube ssh
$ ping host.minikube.internal
PING host.minikube.internal (192.168.39.1): 56 data bytes
```

OK, but try it from inside a pod.
```
$ kubectl run -ti --image=busybox --restart=Never busybox
If you don't see a command prompt, try pressing enter.
/ # ping host.minikube.internal
ping: bad address 'host.minikube.internal'
/ # host host.minikube.internal
sh: host: not found
/ # nslookup host.minikube.internal
Server:		10.96.0.10
Address:	10.96.0.10:53

** server can't find host.minikube.internal: NXDOMAIN

*** Can't find host.minikube.internal: No answer

/ # cat /etc/resolv.conf 
nameserver 10.96.0.10
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
/ # cat /etc/hosts 
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
172.17.0.4	busybox
```

Nothing. OK, I saw that busybox can have issues with DNS lookups, let's try something else:

```
$ kubectl run -ti --image=amazonlinux --restart=Never bash   
If you don't see a command prompt, try pressing enter.
bash-4.2# yum install iputils bind-utils net-tools
[...]
bash-4.2# ping host.minikube.internal
ping: host.minikube.internal: Name or service not known
bash-4.2# nslookup host.minikube.internal
Server:		10.96.0.10
Address:	10.96.0.10#53

** server can't find host.minikube.internal: NXDOMAIN

bash-4.2# cat /etc/resolv.conf 
nameserver 10.96.0.10
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
bash-4.2# cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
172.17.0.4	bash
```

Are my assumptions wrong?

If I knew the IP, it would work -
```
bash-4.2# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.17.0.1      0.0.0.0         UG    0      0        0 eth0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 eth0

bash-4.2# telnet 192.168.39.1 5432
Trying 192.168.39.1...
Connected to 192.168.39.1.
Escape character is '^]'.
```
The port 5432 on host is open and accepting connections.

**Comment by valters** [comment]
Hi @medyagh , I am running minikube on Ubuntu 20.04 with kvm2 driver. Other folks on team are using macOS (Mojave, 10.14.6 ), with VirtualBox driver. (They report less success with HyperKit driver).

The article I am talking about is https://minikube.sigs.k8s.io/docs/handbook/host-access/ , I created this ticket by clicking the link "Create documentation issue" there.

We are running some services on local machine. Postgres is listening on port 5324 on local host. The Postgres port is only one example. We also spin up kafka in a docker container locally, jaeger in a container as a stand-in for full APM collector, etc. These will only exist on local development environment, because in staging/production we will connect to Postgres instance in RDS, or proper Kafka instance in cloud, and so on.

I am having issue getting our service (which we are bringing up in a pod) to talk to Postgres instance on local host. I can get it working by hardcoding the host IP (192.168.39.1 on my machine). I am trying to understand how to make this local configuration generic enough so that all team members could use it - also on macOS with VirtualBox.

My understanding from reading the handbook/host-access/  article is that the DNS name `host.minikube.internal` should be resolvable within a running pod. Could you clarify?

**Comment by jeroenbrouwer** [comment]
Having the exact same problem here! Tried to configure a postgresdb url with `host.minikube.internal` but it only works when using the hardcoded host ip.

**Comment by rahuldhir** [comment]
I have the same use case as @valters. Using GCP CloudSQL in remote k8s environment, but would like to run postgres on the host when deploying to minikube.

Not sure if this is an anti-pattern. If there is a recommended way to do this, it's not clear in the documentation.

**Comment by afbjorklund** [comment]
That minikube article only talks about how to access the host from the VM, **not** from the pods...

It should probably use something like HostAliases for that, to change the `/etc/hosts` file of the pod:

https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/

The misleading sentence seems to be at the top of page: "How to access host resources from a pod"

**Comment by afbjorklund** [comment]
I'm not sure if there is a command to return this address, for use with such commands or specs...

Suppose a workaround is: `minikube ssh grep host.minikube.internal /etc/hosts | cut -f1`

The documentation should probably change the headline to "How to access host resources from VM"

And then add another subsection, on how to access host resources from pods (like the link above).

**Comment by tstromberg** [comment]
This absolutely should work.

My understanding is that the kubelet is supposed to propagate `/etc/hosts` from the VM into the pod at creation time. It appears this may not always be the case.

**Comment by lionelyoung** [comment]
`minikube ssh` and telnet works for me.  Here's my workaround:


1. Get the minikube internal ip
```bash
minikube ssh grep host.minikube.internal /etc/hosts | cut -f1
```
2. Edit your `deployment.yaml` and add this line:
```yaml
...
    spec:
      hostAliases:
      - ip: "192.168.64.1"
        hostnames:
        - "host.minikube.internal"
...
```

**Comment by crucialfelix** [comment]
Neither telnet nor ping are available in the VM:

```
❯ minikube ssh
docker@minikube:~$ telnet
-bash: telnet: command not found
docker@minikube:~$ ping
-bash: ping: command not found
docker@minikube:~$

docker@minikube:~$ cat /etc/*release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
```

So the docs are now out of date I guess.

Inside a pod there is no `host.minikube.internal`

```
cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
172.17.0.8	backend-deployment-5cb8568cd6-f6jt2
```

**Comment by okia** [comment]
Hello!
Also hitting the bug: OSX, minikube 1.14 (Kubernetes 1.19), VirtualBox driver.
If it is a possibility, I'd like to ask to prioritize fixing this bug as it may manifest in the scenarios other than described above and makes it difficult (if possible at all) developing certain Kubernetes scenarios. For example:
Let us suppose we are going to deploy multitier app [onto Kubernetes] and this app will have "external" 
data source (in production that external data source will be, say, RDS instance, but during development
local host' Postgres DB is used). One possible Kubernetes scenario would be to abstract such external
data source behind ExternalName service:
```
apiVersion: v1
kind: Service
metadata:
  name: my-db-service
  namespace: dev
spec:
  type: ExternalName
  externalName: host.minikube.internal
```   
[with the intention to replace `host.minikube.internal` with RDS endpoint in Production]
Now due to this bug, developing/testing this scenario in minikube is somewhat difficult and workaround
described [here](https://github.com/kubernetes/minikube/issues/8439#issuecomment-662244769) won't [directly] work.
In short, it would be nice to fix this bug sooner rather than later :smile:

**Comment by okia** [comment]
@crucialfelix 
> Neither telnet nor ping are available in the VM:

Yeah, noticed that too 😄 
However, nslookup is still there (if I remember correctly). Another way would be to `cat /etc/hosts` while in minikube's shell - you should be able to see host.minikube.internal entry. 

> Inside a pod there is no host.minikube.internal
Indeed, VM's entry is not propagated to pod(s) - that is the essence of this bug.

**Comment by jdart** [comment]
Another use case: when running a php(-fpm) application in a pod we want to connect to an IDE for debugging purposed. To accomplish this today I have to hardcode my wifi interface's ip address which could change when moving between the office and home.

**Comment by juinquok** [comment]
+1 can confirm this is a bug. I am able to curl the MySQL endpoint running on localhost from my minikube ssh but the containers are not able to access it. Have verified this using the testing containers. thanks!

**Comment by rnddude** [comment]
I'm experiencing the same issue. The setup is the following:

```
$ minikube version
minikube version: v1.16.0
commit: 9f1e482427589ff8451c4723b6ba53bb9742fbb1
```
```
minikube start --kubernetes-version=v1.18.6
😄  minikube v1.16.0 on Ubuntu 18.04
🆕  Kubernetes 1.20.0 is now available. If you would like to upgrade, specify: --kubernetes-version=v1.20.0
✨  Using the docker driver based on existing profile
👍  Starting control plane node minikube in cluster minikube
🔄  Restarting existing docker container for "minikube" ...
🐳  Preparing Kubernetes v1.18.6 on Docker 20.10.0 ...
🔎  Verifying Kubernetes components...
🔎  Verifying ingress addon...
🌟  Enabled addons: default-storageclass, storage-provisioner, ingress
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
```
Kubernetes 1.18.6 due to limitations on other Kubernetes environments.

As a workaround I've been using hostAliases, but I would really appreciate a solution to this issue. @tstromberg Is there any fix in sight?

**Comment by jdart** [comment]
> I'm experiencing the same issue. The setup is the following:
> 
> ```
> $ minikube version
> minikube version: v1.16.0
> commit: 9f1e482427589ff8451c4723b6ba53bb9742fbb1
> ```
> 
> ```
> minikube start --kubernetes-version=v1.18.6
> 😄  minikube v1.16.0 on Ubuntu 18.04
> 🆕  Kubernetes 1.20.0 is now available. If you would like to upgrade, specify: --kubernetes-version=v1.20.0
> ✨  Using the docker driver based on existing profile
> 👍  Starting control plane node minikube in cluster minikube
> 🔄  Restarting existing docker container for "minikube" ...
> 🐳  Preparing Kubernetes v1.18.6 on Docker 20.10.0 ...
> 🔎  Verifying Kubernetes components...
> 🔎  Verifying ingress addon...
> 🌟  Enabled addons: default-storageclass, storage-provisioner, ingress
> 🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
> ```
> 
> Kubernetes 1.18.6 due to limitations on other Kubernetes environments.
> 
> As a workaround I've been using hostAliases, but I would really appreciate a solution to this issue. @tstromberg Is there any fix in sight?

Would you mind sharing how your hostAlias works?

**Comment by rnddude** [comment]
> Would you mind sharing how your hostAlias works?

I'm setting hostAliases on my deployments, e.g. 
```
kind: Deployment
apiVersion: apps/v1
metadata:
  name: deployment
spec:
  replicas: 1
  template:
    metadata:
...
    spec:
...
      hostAliases:
        - ip: 192.168.49.1
          hostnames:
            - host.minikube.internal
  strategy:
...
```

The IP is the result of running 
`minikube ssh cat /etc/hosts | grep host.minikube.internal`, which fetches the corresponding entry from the hosts file of the minikube machine.

But you have to set the hostAliases on each and every deployment or whatever you need it on and you'll have to maintain the IP.

**Comment by iamnoah** [comment]
The [workaround](https://github.com/coredns/coredns/issues/2164#issuecomment-427970152) I am using to only have the IP hardcoded in one place is to create a headless service and an endpoint that is the minikube ip:

```yaml

apiVersion: v1
kind: Service
metadata:
  name: minikube-host
  namespace: default
  labels:
    app: minikube-host
spec:
  selector:
    app: minikube-host
  clusterIP: None
  ports:
  - name: app
    port: 8082 # doesn't actually matter

---

kind: Endpoints
apiVersion: v1
metadata:
  name: minikube-host
subsets:
  - addresses:
      - ip: 192.168.64.1 # command below will update this if needed
    ports:
      - port: 8082
        name: app
```

This will update the IP in one go:

```shell
kubectl patch endpoints minikube-host --patch '{"subsets": [{"addresses": [{"ip": "'$( minikube ssh 'grep host.minikube.internal /etc/hosts | cut -f1' | tr -d '[:space:]')'"}]}]}'
```

You can then resolve `minikube-host` (or minikube-host.$namespace.svc.cluster.local).

**Comment by vifar** [comment]
@iamnoah's solution worked for me. I am able to get successful requests.

**Comment by smiletrl** [comment]
@iamnoah 's approach works for me too :)

A little thing though, `selector: app: minikube-host` doesn't seem to be [necessary](https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors). Otherwise the default endpoint will be created.

See terraform code

```
resource "kubernetes_service" "host" {
    metadata {
        name = "minikube-host"

        labels = {
            app = "minikube-host"
        }

        namespace = "default"
    }

    spec {
        port {
            name = "app"
            port = 8082
        }
        cluster_ip = "None"
    }
}

resource "kubernetes_endpoints" "host" {
    metadata {
        name = "minikube-host"
        namespace = "default"
    }
    subset {
        address {
            // This ip comes from command: minikube ssh 'grep host.minikube.internal /etc/hosts | cut -f1'
            ip = "192.168.65.2"
        }

        port {
            name     = "app"
            port     = 8082
        }
    }
}
```

**Comment by Truth1984** [comment]
using `minikube v1.19.0`, then ssh inside 

and `cat /etc/hosts`
```127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
192.168.49.2	minikube
192.168.49.1	host.minikube.internal
192.168.49.2	control-plane.minikube.internal
```
seems like you can use `host.minikube.internal` now without further configuration.

**Comment by denisok** [comment]
I see the same issue, please fix.

**Comment by medyagh** [comment]
This issue is widely been asked by the community and I would like us to either add a workaround to our documentation or fix it

how about we add an test for this and find the root cause ?

**Comment by michaelurban** [comment]
Yes. +1,000,000

**Comment by afbjorklund** [comment]
@medyagh : as far as I know, this is on purpose. Kubernetes only uses the hosts file when using the host network.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L329

**Comment by medyagh** [comment]
@afbjorklund do u mean we should change networking inside minikube to make this work ?

**Comment by medyagh** [comment]
@michaelurban 
do u mind checking if the binary in this PR fixes the issue ?

http://storage.googleapis.com/minikube-builds/11404/minikube-linux-amd64
http://storage.googleapis.com/minikube-builds/11404/minikube-darwin-amd64
http://storage.googleapis.com/minikube-builds/11404/minikube-windows-amd64.exe

**Comment by michaelurban** [comment]
Of course. I should have time later tonight or tomorrow.

**Comment by michaelurban** [comment]
@medyagh No, sorry. I'm not getting anything interesting in my hosts file.

**Comment by cooperaj** [comment]
This doesn't work for standalone docker containers it appears. 

```
> eval $(minikube -p minikube docker-env) 
> docker run --rm -it alpine /bin/sh
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
a0d0a0d46f8b: Already exists 
Digest: sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a
Status: Downloaded newer image for alpine:latest
> ping host.minikube.internal
ping: bad address 'host.minikube.internal'
```

**Comment by sylvain-combe-sonarsource** [comment]
Looking for a solution on this issue for docker containers as well, on MacOS.
An ideal solution would be for host.docker.internal to be resolved, which would allow a smooth transition from docker desktop.

**Comment by sousarmb** [comment]
Tried all solutions above and nothing works. I'm running minikube v1.23.2 with the following scenario:
- Windows 10 Professional
- minikube with 1 node
-- with HyperV driver
-- Docker container runtime
-- Calico CNI

I'm trying to access SSH tunnel to use a DB server in another host. The tunnel is set up in the Windows host with Bitwise and listening on 0.0.0.0 address. 

If i use Docker for Desktop - using host.docker.internal - i can access the DB on the other host without any issues. When i switch to minikube - using host.minikube.internal - i always get a timeout.

I know the Docker for Desktop setup works because i can see connections happenning in the Bitwise log.

**Comment by gorjan-mishevski** [comment]
This is not fixed, I am having the same issues resolving the host inside a container. Mac Monterey

**Comment by Deel96** [comment]
Hi I also face this issue at the moment. Is there an update on this?
In my case I run minikube with the docker driver on an ubuntu 20 host and I want to access a server that runs on 0.0.0.0:4455 on the host from minikube.
Doing
```
minikube ssh
ping host.minikube.internal
```

works but I'm stuck to acess the server that runs on port 4455 on the host.

**Comment by dima-matsola** [comment]
Hello

I face this issue too. My setup:

- Windows 10
    - WLS2 with Ubuntu-18.04
        - Minikube version: v1.25.2
        - Docker version 20.10.12 separately
        
I have an Application A running on the Windows host and listening to port 4441. The URLs param is set like this: `"Urls": "http://*:4441"`. And I have an Application B which should access Application A by this port.

If Application B is started in the minikube in the pod it **cannot** reach Application A by `http://host.minikube.internal:4441` 😢 

Also, if try to test connetion from `minikube ssh` I see that ping actually works:
```
docker@minikube:~$ ping host.minikube.internal
PING host.minikube.internal (192.168.49.1) 56(84) bytes of data.
64 bytes from host.minikube.internal (192.168.49.1): icmp_seq=1 ttl=64 time=0.040 ms
64 bytes from host.minikube.internal (192.168.49.1): icmp_seq=2 ttl=64 time=0.038 ms
64 bytes from host.minikube.internal (192.168.49.1): icmp_seq=3 ttl=64 time=0.069 ms
...
```

but if I test the port it does not work:
```
docker@minikube:~$ nc -vz host.minikube.internal 4441
nc: connect to host.minikube.internal port 4441 (tcp) failed: Connection refused
```
I tried:

- setting URLs as "http://*:4441" or "http://0.0.0.0:4441" or "http://localhost:4441"
- using hostname `host.minikube.internal` or using IP `192.168.49.1`
- even `minikube tunnel` on/off (I gess it should not work with off, but tried anyway, still no result)

However, if Application B is started as a docker container I am able to pass a host name to the container if I grab the IP from `routes -n` on WSL and then add a hostname for it inside docker using `--add-host blabla:<my IP>`. Then the app it is able to reach Application A by `http://blabla:4441` and everything is fine. 

But in minikube, it promises that I can use `host.minikube.internal` without additional actions. Maybe I need to add it to the pod somehow, similar to adding it to the conterner? But I can't modify the helm chart of the Application B.

Please suggest something

Thanks

**Comment by snigdha920** [comment]
I have the exact same issue as @dima-matsola. 

My configuration: 
- Mac OS Monterey v12.5.1, chip: M1 Pro
- minikube version: 1.28.0, commit 986b1ebd987211ed16f8cc10aed7d2c42fc8392f
- docker version: 20.10.21

1. My minikube start command:

```
        minikube start \
          --driver=docker \
          --cpus=4 \
          --memory=7168 \
          --disk-size=50g \
          --kubernetes-version=stable \
          --addons=metrics-server \
          --addons=storage-provisioner \
          --addons=storage-provisioner \
          --addons=ingress
```


2. ping `host.minikube.internal`:

```
PING host.minikube.internal (192.168.65.2) 56(84) bytes of data.
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=1 ttl=37 time=0.995 ms
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=2 ttl=37 time=0.209 ms
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=3 ttl=37 time=0.240 ms
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=4 ttl=37 time=0.234 ms
```

3. ping port NextJS app on port 3002:

```
docker@minikube:~$ nc -vz host.minikube.internal 3002
nc: connect to host.minikube.internal port 3002 (tcp) failed: Connection refused
```

Weird thing is I have a NestJS application running on port 5559, and connection to that port is successful:

```
docker@minikube:~$ nc -vz host.minikube.internal 5559
Connection to host.minikube.internal 5559 port [tcp/*] succeeded!
```


I'm new to minikube and I have no idea how to fix this, any help is appreciated.

**Comment by Fi3** [comment]
Not sure why this is closed I'm using minikube 1.32.0 and host.minikube.internal on the vm solve to the vm ip but on pods it solve to 127.0.0.1. So is completely useless and this guide https://minikube.sigs.k8s.io/docs/handbook/host-access/ is wrong
